ddl/sequence_test.go

```diff
@@ -267,3 +269,379 @@ func (s *testSequenceSuite) TestSequenceAsDefaultValue(c *C) {
+	s.tk.MustQuery("select lastval(seq)").Check(testkit.Rows("10"))
+
+	s.tk.MustExec("drop sequence if exists seq")
+	s.tk.MustExec("create sequence seq increment -1 start 1 maxvalue 10 minvalue -10 cache 3 cycle")
+	s.tk.MustQuery("select nextval(seq)").Check(testkit.Rows("1"))
+	s.tk.MustQuery("select setval(seq, -8)").Check(testkit.Rows("-8"))
+	s.tk.MustQuery("select nextval(seq)").Check(testkit.Rows("-9"))
+	sequenceTable = testGetTableByName(c, s.tk.Se, "test", "seq")
+	tc, ok = sequenceTable.(*tables.TableCommon)
+	c.Assert(ok, Equals, true)
+	_, end, round = tc.GetSequenceCommon().GetSequenceBaseEndRound()
+	c.Assert(end, Equals, int64(-10))
+	c.Assert(round, Equals, int64(0))
+}
+
+func (s *testSequenceSuite) TestInsertSequence(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use test")
+	s.tk.MustExec("drop sequence if exists seq")
+	s.tk.MustExec("drop table if exists t")
+
+	// test insert with sequence default value.
+	s.tk.MustExec("create sequence seq")
+	s.tk.MustExec("create table t (a int default next value for seq)")
+	s.tk.MustExec("insert into t values()")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("1"))
+	s.tk.MustExec("insert into t values(),(),()")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("1", "2", "3", "4"))
+	s.tk.MustExec("delete from t")
+	s.tk.MustExec("insert into t values(-1),(default),(-1)")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("-1", "5", "-1"))
+
+	// test insert with specified sequence value rather than default.
+	s.tk.MustExec("drop table if exists t")
+	s.tk.MustExec("create table t (a int)")
+	s.tk.MustExec("insert into t values(next value for seq)")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("6"))
+	s.tk.MustExec("insert into t values(next value for seq),(nextval(seq))")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("6", "7", "8"))
+
+	// test insert with sequence expression.
+	s.tk.MustExec("delete from t")
+	s.tk.MustExec("insert into t values(next value for seq + 1),(nextval(seq) * 2)")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("10", "20"))
+	s.tk.MustExec("delete from t")
+	s.tk.MustExec("insert into t values((next value for seq - 1) / 2)")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("5"))
+
+	// test insert with user specified value.
+	s.tk.MustExec("delete from t")
+	s.tk.MustExec("insert into t values(-1),(next value for seq),(-1),(nextval(seq))")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("-1", "12", "-1", "13"))
+
+	// test insert with lastval & setval.
+	s.tk.MustExec("delete from t")
+	s.tk.MustExec("insert into t values(lastval(seq)),(-1),(nextval(seq))")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("13", "-1", "14"))
+	s.tk.MustExec("delete from t")
+	s.tk.MustExec("select setval(seq, 100)")
+	s.tk.MustExec("insert into t values(lastval(seq)),(-1),(nextval(seq))")
+	s.tk.MustQuery("select * from t").Check(testkit.Rows("14", "-1", "101"))
+}
```

Comment:
Add some test for using `nextval()`, `setval()`, `lastval()` in generated columns' expression.

Reply:
We store sequence value by int64. Could you add some tests for integer overflow?

Reply:
nice suggestion.